shortcut.go

```diff
@@ -46,30 +48,46 @@ func New(ipv4Subnets []string, ipv6Subnets []string) Shortcut {
 		len(ipv6Subnets),
 	)
 	return &shortcut{
-		v4list: newSortList(ipv4Subnets),
-		v6list: newSortList(ipv6Subnets),
-		// Prefers the system resolver by default, hopefully can use OS DNS cache.
-		resolver: net.DefaultResolver,
+		v4list:   newSortList(ipv4Subnets),
+		v6list:   newSortList(ipv6Subnets),
+		resolver: defaultResolver,
 	}
 }
 
-func (s *shortcut) Allow(ctx context.Context, addr string) (bool, net.IP) {
+func defaultResolver(ctx context.Context, addr string) (net.IP, error) {
 	host, _, err := net.SplitHostPort(addr)
 	if err != nil {
 		host = addr
 	}
-	addrs, err := s.resolver.LookupIPAddr(ctx, host)
+	addrs, err := net.DefaultResolver.LookupIPAddr(ctx, host)
 	if err != nil {
-		return false, nil
+		return nil, err
 	}
 	for _, addr := range addrs {
-		ip := addr.IP.To4()
-		if ip != nil {
-			return s.v4list.Contains(ip), ip
+		if ip := addr.IP.To4(); ip != nil {
+			return ip, nil
 		}
-		if ip = ip.To16(); ip != nil {
```

Comment:
There was a bug. If `ip` is nil, `ip.To16()` is `nil` as well (fortunately it doesn't panic).